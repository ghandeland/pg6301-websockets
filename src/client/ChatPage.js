import React, { useState, useEffect, useRef } from 'react'

const useWebSocketChatPage = () => {
    const [chatLog, setChatLog] = useState([]);
    const [ws, setWs] = useState();
    const connected = useRef(false)
    
    const connect = () => {
      console.log("Connecting");
      const ws = new WebSocket("ws://" + window.location.host);
      setWs(ws);
      ws.onopen = (event) => {
        console.log("Opened", event);
        connected.current = true;
      };
      ws.onclose = (event) => {
        console.log('Closed', event);
        connected.current = false;
      };
      ws.onmessage = (msg) => {
        const message = JSON.parse(msg.data);
        setChatLog((chatLog) => [...chatLog, message]);
      }
    }
    
    useEffect(() => connect(), []);
    
    const handleSendMessage = (msg) => {
      ws.send(JSON.stringify(msg));
    };
    return { chatLog, handleSendMessage };
}


export const ChatPage = () => {
  const { chatLog, handleSendMessage } = useWebSocketChatPage();

  return <ChatView chatLog={chatLog} onSendMessage={handleSendMessage} />;
  
}

export const ChatView = ({chatLog, onSendMessage}) => {
  const [message, setMessage] = useState('');

  const handleSubmit = (e) => {
    e.preventDefault();
    onSendMessage(message);
    setMessage('');
  };

  return (
    <div className="chat-container">
      <header>
        <h1>Chat page</h1>
      </header>
      <main>
        <div className="chat-display">
          {chatLog.map((message, index) => {
            return <div className="chat-msg" key={index}>{message}</div>;
          })}
        </div>
      </main>
      <footer>
        <form onSubmit={handleSubmit}>
          <input
            type="text"
            className="chat-input"
            onChange={(e) => setMessage(e.target.value)}
            autoFocus={true}
            value={message}
          />
          <button value="Enter">Enter</button>
        </form>
      </footer>
    </div>
  );
}

export default ChatPage;