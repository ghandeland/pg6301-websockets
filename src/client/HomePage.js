import React from 'react'

const HomePage = () => {
    return (
      <div>
        <h1>Websockets exercise</h1>
        <p>Exercise using websockets with node and express.</p>
      </div>
    );
}


export default HomePage
