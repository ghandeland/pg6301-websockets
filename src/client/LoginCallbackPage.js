import React, { useEffect } from 'react'
import { useHistory } from 'react-router';

const LoginCallbackPage = ({ onAccessToken }) => {
    
    const history = useHistory();
    
    const hash = Object.fromEntries(
        new URLSearchParams(window.location.hash.substr(1))
    );
    console.log("hash", hash);
    
    useEffect(() => {
        onAccessToken(hash.access_token);
        history.push("/");
    }, [])
    
    return (
        <div>
            <h1>Login Callback Page</h1>      
        </div>
    )
}

export default LoginCallbackPage
