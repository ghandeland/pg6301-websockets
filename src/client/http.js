export async function fetchJson(url) {
  const res = await fetch(url);
  if (!res.ok) {
    console.log('Hello');
    throw new Error(`Failed to fetch ${url}: ${res.status} ${res.statusText}`);
  }
  const json = await res.json();
  return json;
}

exports.fetchJson = fetchJson;