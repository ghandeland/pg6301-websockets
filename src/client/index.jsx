import React from 'react';
import ReactDOM from 'react-dom';

const index = () => {
    return (
        <div>
            <h1>Hello world</h1>
        </div>    
    )
}

ReactDOM.render(index, document.getElementById("root"));