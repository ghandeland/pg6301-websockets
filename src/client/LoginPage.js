import React from 'react';
import { fetchJson } from './http';

const LoginPage = ({ identityProvider }) => {
  const { discoveryUrl, client_id, scope } = identityProvider;

  const handleLogin = async () => {
    const { authorization_endpoint } = await fetchJson(discoveryUrl);
    const params = {
      response_type: "token",
      client_id,
      scope,
      redirect_uri: window.location.origin + "/login/callback"
    };
    window.location.href = authorization_endpoint + "?" + new URLSearchParams(params);
    
  }
  
  return (
    <div>
      <h1>Login page</h1>
      <button onClick={() => handleLogin()}>Log in</button>
    </div>
  );
};

export default LoginPage;
