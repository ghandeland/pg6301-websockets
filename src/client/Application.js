import React, { useState } from 'react';
import ReactDOM from 'react-dom';
import { BrowserRouter, Switch, Route } from 'react-router-dom';
import HomePage from './HomePage';
import NavBar from './NavBar';
import ChatPage from './ChatPage';
import ProfilePage from './ProfilePage';
import LoginPage from './LoginPage';
import fetchJson from './http';

const Application = () => {
  const [accessToken, useAccessToken] = useState();

  const loadProfile = () => {
    const at = accessToken ? { Authorization: 'Bearer ' + accessToken } : {};
    fetchJson('/api/profile', {
      headers: { ...at },
    });
  };

  const googleIdentityProvider = {
    discoveryUrl: process.env.REACT_APP_GOOGLE_OAUTH_DISCOVERY_URL,
    client_id: process.env.REACT_APP_GOOGLE_OAUTH_CLIENT_ID,
    scope: 'openid profile email',
  };
  
  return (
    <div>
      <BrowserRouter>
        <header>
          <NavBar />
        </header>
        <Switch>
          <Route exact path="/">
            <HomePage />
          </Route>
          <Route exact path="/chat">
            <ChatPage />
          </Route>
          <Route exact path="/profile">
            <ProfilePage />
          </Route>
          <Route exact path="/login">
            <LoginPage identityProvider={googleIdentityProvider} />
          </Route>
          <Route exact path="/login/callback">
            <h1>Login callback</h1>
          </Route>
        </Switch>
      </BrowserRouter>
    </div>
  );
};

export default Application;
