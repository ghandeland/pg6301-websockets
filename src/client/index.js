import React from 'react';
import ReactDOM from 'react-dom';
import { BrowserRouter, Switch, Route, Link } from 'react-router-dom'
import HomePage from './HomePage'
import ChatPage from './ChatPage';
import ProfilePage from './ProfilePage';
import LoginPage from './LoginPage';

const App = () => {
    return (
      <div>
        <header></header>
        <BrowserRouter>
          <Link to="/">Homepage</Link>
          {' | '}
          <Link to="/chat">Chat</Link>
          {' | '}
          <Link to="/profile">Profile</Link>
          {' | '}
          <Link to="/login">Login</Link>
          <Switch>
            <Route exact path="/">
              <HomePage />
            </Route>
            <Route exact path="/chat">
              <ChatPage />
            </Route>
            <Route exact path="/profile">
              <ProfilePage />
            </Route>
            <Route exact path="/login">
              <LoginPage />
            </Route>
          </Switch>
        </BrowserRouter>
      </div>
    );
}

ReactDOM.render(<App/>, document.getElementById("root"));