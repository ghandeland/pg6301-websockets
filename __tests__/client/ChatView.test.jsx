import { ChatView } from "../../src/client/ChatPage";
import React from 'react';
import ReactDOM from 'react-dom';

describe("chat view", () => {
    it("shows chat log as messages", () => {
        
        const cLog = ['hello', 'foo', "bar"];
        const handleSendMsg = jest.fn()
        
        const container = document.createElement("div");
        ReactDOM.render(
          <ChatView chatLog={cLog} onSendMessage={handleSendMsg} />,
          container
        );
                
        expect(
          container.querySelector('.chat-msg')
            .textContent
        ).toEqual('hello');
        expect(container).toMatchSnapshot();
    })
    
});