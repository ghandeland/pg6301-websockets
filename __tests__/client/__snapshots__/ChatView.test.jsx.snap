// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chat view shows chat log as messages 1`] = `
<div>
  <div
    class="chat-container"
  >
    <header>
      <h1>
        Chat page
      </h1>
    </header>
    <main>
      <div
        class="chat-display"
      >
        <div
          class="chat-msg"
        >
          hello
        </div>
        <div
          class="chat-msg"
        >
          foo
        </div>
        <div
          class="chat-msg"
        >
          bar
        </div>
      </div>
    </main>
    <footer>
      <form>
        <input
          class="chat-input"
          type="text"
          value=""
        />
        <button
          value="Enter"
        >
          Enter
        </button>
      </form>
    </footer>
  </div>
</div>
`;
